### PR TITLE
fix: Add stricter check for "use server" exports

### LINKED ISSUE #62727: 14.1.1 version introducing breaking change for cookies()
### Link to the code that reproduces this issue

https://github.com/alimek/next-bug

### To Reproduce

```
export const getToken = cache(() => {
  const cookieStore = cookies();
  const token = cookieStore.get("token")?.value;

  return token ?? 'no-token';
});

export const getLanguage = cache(() => {
  const cookieStore = cookies();
  const language = cookieStore.get("language")?.value;

  return language ?? 'pl-pl';
});
```

when calling getLanguage() or getToken() it should return always string, after upgrade to 14.1.1 its Promise 

### Current vs. Expected behavior

Current

 ```
export default async function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
 const token = getToken();
 const language = getLanguage();

  return (
    <html lang={language} suppressHydrationWarning>
      <body className={inter.className} suppressHydrationWarning>
        <Client value={token} />
      </body>
    </html>
  );
}
```

generating 

```html
<html lang="[object Promise]">
...
</html>
```
instead

```html
<html lang="pl-pl">
...
</html>
```

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.3.0: Wed Dec 20 21:30:44 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T6000
Binaries:
  Node: 18.19.0
  npm: 10.2.3
  Yarn: 4.1.0
  pnpm: 8.7.0
Relevant Packages:
  next: 14.1.1
  eslint-config-next: 14.1.0
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: standalone
```

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next dev (local), next build (local)

### Additional context

_No response_

<sub>[NEXT-2660](https://linear.app/vercel/issue/NEXT-2660/1411-version-introducing-breaking-change-for-cookies)</sub>

### PR DESCRIPTION
**Pull request by shuding**
As mentioned in the new-added error messages, and the [linked resources](https://react.dev/reference/react/use-server#:~:text=Because%20the%20underlying%20network%20calls%20are%20always%20asynchronous%2C%20%27use%20server%27%20can%20only%20be%20used%20on%20async%20functions.):

> Because the underlying network calls are always asynchronous, 'use server' can only be used on async functions.
> https://react.dev/reference/react/use-server

It's a requirement that only async functions are allowed to be exported and annotated with `'use server'`. Currently, we already have compiler check so this will already error:

```js
'use server'

export function foo () {} // missing async
```

However, since exported values can be very dynamic the compiler can't catch all mistakes like that. We also have a runtime check for all exports in a `'use server'` function, but it only covers `typeof value === 'function'`.

This PR adds a stricter check for "use server" annotated values to also make sure they're async functions (`value.constructor.name === 'AsyncFunction'`).

That said, there are still cases like synchronously returning a promise to make a function "async", but it's still very different by definition. For example:

```js
const f = async () => { throw 1; return 1 }
const g = () => { throw 1; return Promise.resolve(1) }
```

Where `g()` can be synchronously caught (`try { g() } catch {}`) but `f()` can't even if they have the same types. If we allow `g` to be a Server Action, this behavior is no longer always true but depending on where it's called (server or client).

Closes #62727.

### CODE REVIEW

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 6d1476ff1206a2cf18f91bd31eb8b811e0230e81

`pnpm test-dev test/e2e/app-dir/metadata/metadata.test.ts` (PPR)
- app dir - metadata > react cache > should have same title and page value when navigating

<details>
<summary>Expand output</summary>

  ● app dir - metadata › react cache › should have same title and page value when navigating

    expect(received).toBe(expected) // Object.is equality

    Expected: "0.22845829255664696"
    Received: "0.09518457533937363"

      973 |         const obj = JSON.parse(title)
      974 |         // Check `cache()`
    > 975 |         expect(obj.val.toString()).toBe(value)
          |                                    ^
      976 |         // Check `fetch()`
      977 |         // TODO-APP: Investigate why fetch deduping doesn't apply but cache() does.
      978 |         if (!isNextDev) {

      at Object.toBe (e2e/app-dir/metadata/metadata.test.ts:975:36)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| buildDuration | 17.3s | 17.3s | N/A |
| buildDurationCached | 9.4s | 8.1s | N/A |
| nodeModulesSize | 198 MB | 198 MB | ⚠️ +3.9 kB |
| nextStartRea..uration (ms) | 507ms | 500ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.2 kB | 30.2 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 243 B | 241 B | N/A |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 45.4 kB | 45.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.2 kB | 4.2 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.56 kB | 6.56 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| index.html gzip | 529 B | 529 B | ✓ |
| link.html gzip | 543 B | 541 B | N/A |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 529 B | 529 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95 kB | 95 kB | N/A |
| page.js gzip | 3.08 kB | 3.08 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 624 B | ✓ |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 1.61 kB | 1.61 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 96.7 kB | 96.7 kB | ✓ |
| app-page-tur..prod.js gzip | 98.5 kB | 98.5 kB | ✓ |
| app-page-tur..prod.js gzip | 92.9 kB | 92.9 kB | ✓ |
| app-page.run...dev.js gzip | 150 kB | 150 kB | ✓ |
| app-page.run..prod.js gzip | 91.4 kB | 91.4 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 20.9 kB | 20.9 kB | ✓ |
| app-route.ru..prod.js gzip | 14.7 kB | 14.7 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.51 kB | 9.51 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.79 kB | 9.79 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.51 kB | 9.51 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 23 kB | 23 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.6 kB | 50.6 kB | ✓ |
| Overall change | 950 kB | 950 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong></summary>

|  | vercel/next.js canary  | vercel/next.js shu/092d | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | N/A |
| index.pack gzip | 106 kB | 106 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 6d1476ff1206a2cf18f91bd31eb8b811e0230e81

**Review by huozhi — APPROVED** [review]

**Review by ztanner — APPROVED** [review]

**Review by ijjk — APPROVED** [review]

**Comment by sannajammeh** [comment]
This immediately breaks `cache` function from React.

**Comment by danieltott** [comment]
After updating to `14.1.2` we got surprised by an issue - we have an async server action function in a file marked with `use server`. That function was calling another function in the same file that was not async, but was exported (turns out for no reason, not using it anywhere else). I didn't get any warnings from nextjs on this, but it absolutely broke the original function since the other fn was now returning a promise instead of it's original value. Removing `export` from the function definition fixed that - but this was certainly unexpected.

The new fix seems to be automatically converting all exported functions in a `use server` file to async with no warning. Is this like...a bug? As in, should I log it somewhere? It certainly feels like a breaking change (since it did break our production build), but I'm not sure what to do with that information. If "do nothing" is the answer that's fine, I just wanted to bring it up.

**Comment by mertafor** [comment]
> After updating to `14.1.2` we got surprised by an issue - we have an async server action function in a file marked with `use server`. That function was calling another function in the same file that was not async, but was exported (turns out for no reason, not using it anywhere else). I didn't get any warnings from nextjs on this, but it absolutely broke the original function since the other fn was now returning a promise instead of it's original value. Removing `export` from the function definition fixed that - but this was certainly unexpected.
> 
> The new fix seems to be automatically converting all exported functions in a `use server` file to async with no warning. Is this like...a bug? As in, should I log it somewhere? It certainly feels like a breaking change (since it did break our production build), but I'm not sure what to do with that information. If "do nothing" is the answer that's fine, I just wanted to bring it up.

Same here. I noticed the issue by an unexpected returned value, otherwise there was no errors or warnings. I was able to debug the issue by logging the result returned by server action and realize that it actually returns a promise even though the function is synchronous. Apparently all functions placed in a "use server" file are automatically turned into async functions, even though definitions are sync. In my opinion it was the missing piece for use server files and definitely a great improvement. 

But we need an Eslint check/error for sync functions placed in use server files. For example I also use a sync function for cache() in the same file without exporting and I don't know if it's the right behavior at all.

File: packages/next/src/build/webpack/loaders/next-flight-loader/action-validate.ts [hunk 1]
```diff
@@ -6,7 +6,22 @@ export function ensureServerEntryExports(actions: any[]) {
...
       throw new Error(
-        `A "use server" file can only export async functions, found ${typeof action}.`
+        `A "use server" file can only export async functions, found ${typeof action}.\nRead more: https://nextjs.org/docs/messages/invalid-use-server-value`
+      )
+    }
+
+    if (action.constructor.name !== 'AsyncFunction') {
+      const actionName: string = action.name || ''
+
+      // Note: if it's from library code with `async` being transpiled to returning a promise,
```
Comment (chungweileong94) [inline packages/next/src/build/webpack/loaders/next-flight-loader/action-validate.ts:16]: This bit is actually quite annoying for library authors tho, any plan where this will get improved in the future?

Comment (sam3d) [reply packages/next/src/build/webpack/loaders/next-flight-loader/action-validate.ts:16]: Agree. It also breaks the use of Zod for validating the input of Server Actions (this worked before):

```ts
"use server";

import { db } from "./db";
import { z } from "zod";

export const getUserName = z
  .function()
  .args(z.object({ userId: z.string() }))
  .implement(async ({ userId }) => {
    const user = await db.users.query({ where: { id: userId } });
    return user.name;
  });
```

**Comment by shuding** [comment]
Hey everyone, thanks for all your feedback!

## Tl;DR

`"use server"` **requires** the functions to be async. If they are not, they will be converted to async functions.

So, instead of silently converting your sync functions to be async (breaks type checking), here we're directly showing errors and ask you to make them async in the first place.

## Details

> it breaks `React.cache`

Please give https://github.com/vercel/next.js/issues/62860#issuecomment-1978730839 a read - `React.cache` will not do anything when it's called inside a Server Action.

> but it absolutely broke the original function since the other fn was now returning a promise instead of it's original value

As per this PR description, async functions and sync functions that return promises are fundamentally different things:

```js
const f = async () => { throw 1; return 1 }
const g = () => { throw 1; return Promise.resolve(1) }
```

> The new fix seems to be automatically converting all exported functions in a use server file to async with no warning. Is this like...a bug?

Again, as mentioned in the PR description, "Because the underlying network calls are always asynchronous, 'use server' can only be used on async functions." (https://react.dev/reference/react/use-server) that's a **requirement**. Previously that's not ensured which was a bug and this PR fixes it.

> But we need an Eslint check/error for sync functions placed in use server files.

This would be great, but it's difficult to ensure the type statically. Also in TS `async () => {}` and `() => Promise` are considered the same type but they're actually different things.

> Agree. It also breaks the use of Zod for validating the input of Server Actions (this worked before)

Agreed that this is annoying. I'll see if there are any other work arounds for it.
